test/Application/MarshalMiddlewareTraitTest.php

```diff
@@ -0,0 +1,298 @@
+        $this->assertAttributeSame($base, 'middleware', $middleware);
+        $this->assertAttributeSame($this->responsePrototype->reveal(), 'responsePrototype', $middleware);
+    }
+
+    public function testPreparingArrayOfMiddlewareReturnsMiddlewarePipe()
+    {
+        $first  = $this->prophesize(ServerMiddlewareInterface::class)->reveal();
+        $second = function ($request, DelegateInterface $delegate) {
+        };
+        $third  = function ($request, $response, callable $next) {
+        };
+        $fourth = 'fourth';
+        $fifth  = TestAsset\CallableMiddleware::class;
+        $sixth  = TestAsset\CallableInteropMiddleware::class;
+
+        $this->container->has('fourth')->willReturn(true);
+        $this->container->has(TestAsset\CallableMiddleware::class)->willReturn(false);
+        $this->container->has(TestAsset\CallableInteropMiddleware::class)->willReturn(false);
+
+        $base = [
+            $first,
+            $second,
+            $third,
+            $fourth,
+            $fifth,
+            $sixth,
+        ];
+
+        $middleware = $this->prepareMiddleware($base);
+        $this->assertInstanceOf(MiddlewarePipe::class, $middleware);
+
+        $r = new ReflectionProperty($middleware, 'pipeline');
+        $r->setAccessible(true);
+        $pipeline = $r->getValue($middleware);
+
+        $this->assertCount(6, $pipeline);
+    }
+
+    public function testPreparingArrayOfMiddlewareWithoutContainerReturnsMiddlewarePipe()
+    {
+        $first  = $this->prophesize(ServerMiddlewareInterface::class)->reveal();
+        $second = function ($request, DelegateInterface $delegate) {
+        };
+        $third  = function ($request, $response, callable $next) {
+        };
+        $fifth  = TestAsset\CallableMiddleware::class;
+        $sixth  = TestAsset\CallableInteropMiddleware::class;
+
+        $base = [
+            $first,
+            $second,
+            $third,
+            $fifth,
+            $sixth,
+        ];
+
+        $middleware = $this->prepareMiddleware($base);
+        $this->assertInstanceOf(MiddlewarePipe::class, $middleware);
+
+        $r = new ReflectionProperty($middleware, 'pipeline');
+        $r->setAccessible(true);
+        $pipeline = $r->getValue($middleware);
+
+        $this->assertCount(5, $pipeline);
+    }
+
+    public function testPreparingArrayOfMiddlewareRaisesExceptionWhenContainerMissingAndServiceInvalid()
+    {
+        $first  = $this->prophesize(ServerMiddlewareInterface::class)->reveal();
+        $second = 'second-middleware';
+        $third  = 'third-middleware';
+
+        // No container is passed to the method, so this will not matter
+        $this->container->has('second-middleware')->willReturn(true);
+        $this->container->has('third-middleware')->willReturn(true);
+
+        $base = [$first, $second, $third];
+
+        $this->setExpectedException(InvalidMiddlewareException::class, 'second-middleware');
```

Comment:
As I can see we are using PHPUnit 5.7, so here we should use:
`expectException` and `expectExceptionMessage` instead of deprecated method `setExpectedException`.